### PR TITLE
Emit multiple fields from the same runtime field script

### LINKED ISSUE #68203: Add support for emitting multiple fields values from a script
We recently added support for runtime fields, that are computed at search time based on a painless script. As of today, a runtime field script can emit values for a single field, the one that the script is declared under.

We would like to add the ability for a script to emit values for multiple fields. This will be achieved by introducing support for a new field type (name to be defined) as part of the runtime section. Its script emits fields that belong to such object. This is  particularly useful given that scripts support grok and dissect (#68088):

```
PUT localhost:9200/logs/_mappings
{
  "runtime" : {
    "log" : {
      "type" : "tbd",
      "script": '''
        emit(grok('%{COMMONAPACHELOG}').extract(doc["message"].value)));
      ''',
      "fields" : {
        "clientip" : {
          "type" : "ip"
        },
        "verb" : {
          "type" : "keyword"
        },
        "request" : {
          "type" : "keyword"
        },
        "response" : {
          "type" : "long"
        }
      }
    }
  },
  "properties" : {
    "message" : {
      "type" : "keyword"
    }
  }
}
```

In the example above, the `grok` function splits the message field into sub-fields based on the provided grok pattern, and each of the resulting fields is emitted in the following loop. The emitted fields need to be listed under the sub-fields in order to specify their type and make them searchable (and discoverable through field_caps) like any other field:

```
POST /logs*/_search
{
  "aggs": {
    "response_codes": {
      "range": {
        "field": "log.response",
        "ranges": [
          { "to": 300 },
          { "from": 300, "to": 400 },
          { "from": 500 }
        ]
      }
    }
  }
}
```

### PR DESCRIPTION
**Pull request by javanna**
We have recently introduced support for grok and dissect to the runtime fields Painless context that allows to split a field into multiple fields. Though each runtime field can only emit values for a single field. This PR introduces support for emitting multiple fields from the same script.

Note that this is a draft to share the path forward and gather initial feedback. We have not settled yet on the API. I have introduced support for an object field type on the runtime section, with a new script context that accepts two emit signatures:
- emit(String, Object)
- emit(Map)

We may use the object type, or introduce a new field type.

The way that it emits multiple fields is by returning multiple `MappedFieldType`s from `RuntimeField#asMappedFieldTypes`. The sub-fields are instances of the runtime fields that are already supported, with a little tweak to adapt the script defined by their parent to an artificial script factory for each of the sub-fields that makes the relevant sub-field accessible. This approach allows to reuse all of the existing runtime fields code for the sub-fields.

Once we settled on the API, besides addressing the TODOs, we will need to add support for making the new field type indexed by moving its definition to the properties section.

Closes #68203

### CODE REVIEW

**Review by rjernst — COMMENTED** [review]
A couple thoughts

File: modules/lang-painless/src/main/resources/org/elasticsearch/painless/spi/org.elasticsearch.script.object_field.txt [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+# or more contributor license agreements. Licensed under the Elastic License
+# 2.0 and the Server Side Public License, v 1; you may not use this file except
+# in compliance with, at your election, the Elastic License 2.0 or the Server
+# Side Public License, v 1.
+#
+
+# The whitelist for object runtime fields
+
+# These two whitelists are required for painless to find the classes
+class org.elasticsearch.script.ObjectFieldScript @no_import {
```
[the remark this answers is not in the document]
Comment (rjernst) [reply modules/lang-painless/src/main/resources/org/elasticsearch/painless/spi/org.elasticsearch.script.object_field.txt:12]: This type is the script itself right? It's the "this"? It seems odd to need to allow that; a user shouldn't need to know about it, and that is reinforced by the fact no methods are allowed on it, only the type itself is allowed here.

[the remark this answers is not in the document]
Comment (jdconrad) [reply modules/lang-painless/src/main/resources/org/elasticsearch/painless/spi/org.elasticsearch.script.object_field.txt:12]: You are correct about the type is the script itself. We have no special knowledge of anything in the whitelist loading. I'm happy to consider possible ways to add specialized code for this, but for now, this is the correct way to whitelist this structure.

File: server/src/main/java/org/elasticsearch/script/ObjectFieldScript.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+     */
+    public final List<Object> getValues(String field) {
+        //TODO for now we re-run the script every time a leaf field is accessed, but we could cache the values?
+        fieldValues.clear();
+        execute();
+        return fieldValues.get(field);
+    }
+
+    protected final void emit(String field, Object value) {
+        List<Object> values = this.fieldValues.computeIfAbsent(field, s -> new ArrayList<>());
```
[the remark this answers is not in the document]
Comment (javanna) [reply server/src/main/java/org/elasticsearch/script/ObjectFieldScript.java:54]: Given that in most cases we end up trying to parse what `toString` returns when the type is not exactly what we'd like, I am not sure how we would restrict the type of the argument that is provided here. That is maybe a consequence of re-using the existing code for parsing from _source that tries to adapt to the different field types that may be found in _source.

**Comment by javanna** [comment]
thanks @jdconrad 

> Looking at the code I wonder if would make sense to just have two emit methods per type where one is emit(Object) -> emit(def) and the other is emit(String, Object) -> emit(String, def). This would allow the emitToObject methods you have to replace emit and get the conversions we want anyway. I wonder how bad boxing really is on the perf hit. In a number of cases it may already be boxed depending on what they do with the values

I was thinking the same, maybe that is a potential follow-up. I think it deserves more discussion, the two concerns I could see is that we would introduce more leniency in the existing runtime field types, and once we've done that we will hardly have a way to go back. Another concern could be the auto-boxing performance costs, like you said. A big advantage though would be to help users tremendously with their type conversions, which I am sure they struggle with at the moment.

**Comment by javanna** [comment]
Closing this, we will work in a feature branch for the last couple of things that need to be completed and open another PR when ready. Thanks for the early feedback everybody.

**Pull request closed without merging**